fix: utilize dd-trace's provided scope management

dowster opened the pull request:
## :memo: Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

dd-trace@1.5.1 was not receiving the scope properly with the code in place. Any auto-instrumented traces dd-trace generated would have unique trace IDs and no parent links. This change will fix the scope handling. Like the existing implementation, this will only work with [dd-trace@1.5.1 `async_hooks` scope management](https://github.com/DataDog/dd-trace-js/blob/v1.5.1/packages/dd-trace/src/scope/async_hooks.js).

@ngraef had made a similar patch with PR #895. He also included support for other scope managers. I'd rather not implement that now as we're looking at working with OpenTelemetry libraries for further tracing enhancements and possibly switching from dd-trace.


### :dart: Relevant issues
<!-- Please add relevant opened issues -->

### :gem: Type of change

<!-- Please delete options that are not relevant. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### :scroll: Example code
```js

``` 

## :vertical_traffic_light: How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

This change was deployed to our testing environment for a few different services. I verified that trace propagation was working correctly with dd-trace's Postgres, Redis, HTTP, TCP, and file system instrumentation. 

We are using the dd-trace@1.5.1 library.

## :checkered_flag: Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] **I have added tests that prove my fix is effective or that my feature works**
- [x] **New and existing unit tests pass locally with my changes**
- [ ] I have commented my code, particularly in hard-to-understand areas

OutdatedVersion commented on src/tracing/exporters/datadog.js:
```diff
@@ -157,17 +157,10 @@ class DatadogTraceExporter extends BaseTraceExporter {
-		this.ddScope._spans[asyncId] = ddSpan;
+		this.ddTracer.scope()._enter(ddSpan);
 
 		span.meta.datadog = {
```
Any idea on if this meta property is considered internal? I could see the `asyncId` being useful in some middlewares

icebob commented:
How this changes work with older versions? Because the Moleculer 0.14 supports the following dd-trace versions: `"dd-trace": "^0.33.0 || ^0.34.0 || ^0.35.0 || ^0.36.0 || >=1.0.0 <1.6.0",`

If it doesn't work with older and we should unsupport them, it means, this PR can be released only in the next major 0.15 version of Moleculer

dowster commented:
> How this changes work with older versions? Because the Moleculer 0.14 supports the following dd-trace versions: `"dd-trace": "^0.33.0 || ^0.34.0 || ^0.35.0 || ^0.36.0 || >=1.0.0 <1.6.0",`
> 
> If it doesn't work with older and we should unsupport them, it means, this PR can be released only in the next major 0.15 version of Moleculer

I can add a check my change to utilize existing logic if dd-trace is less than 1.5.1, and utilize my change if it is 1.5.1 since I have tested on 1.5.1. 

On the other hand, could we push this change out with version 0.15.0 and remove support for older dd-trace versions? I think allowing more breaking changes / frequent version bumps could possibly help move the project along more quickly.

icebob commented:
Yeah, this is the plan. For this purpose, could you change this PR with removing old dd-trace versions and change the target branch from master to `next`, please?

dowster commented:
I rebased the PR onto the `next` branch, updated the peerDependency and devDependency of datadog to the latest major, and updated a test which broke due to a change inside dd-trace. 

I figured going to dd-trace@4 and dropping all others should be acceptable as older versions of dd-trace support node@12 and node@14 while both moleculer/next and dd-trace@4 require node>=16. 

I still need to do some integration testing with this build in my cluster to confirm nothing breaks. This version of dd-trace includes [moleculer integration](https://github.com/DataDog/dd-trace-js/tree/master/packages/datadog-plugin-moleculer), so I want to be sure we don't start seeing double counted spans. In a future change, we may be able to remove most of the exporter logic in the moleculer trace exporter and rely completely on dd-trace's instrumentation.

icebob commented:
>  In a future change, we may be able to remove most of the exporter logic in the moleculer trace exporter and rely completely on dd-trace's instrumentation.

In the future I will remove the built-in tracing solution and using only OpenTelemetry: https://github.com/moleculerjs/moleculer/discussions/1125

icebob commented:
@dowster any news about it? I'm working on to update all dependencies and dd-trace is the latest one but it conflicts with your PR. I've checked your PR and I've got error messages. Could you work on it to fix it and I can merge it to the next branch?

dowster commented:
> @dowster any news about it? I'm working on to update all dependencies and dd-trace is the latest one but it conflicts with your PR. I've checked your PR and I've got error messages. Could you work on it to fix it and I can merge it to the next branch?

I might not be able to get back to this for another week. I'll also need to set up some new test scenarios, since I won't have access to my current test environments or datadog accounts. I'll post back here later to update if it's still feasible for me to complete this.

icebob commented:
Plz post back if you are not feasible because in this case, I should work on it.
